src/hooks/useGlobalIntersectionObserverWithDelay.ts

```diff
@@ -0,0 +1,125 @@
+ * globally available. Once it's created, it can be used by any element that would like
+ * to be observed using that observer using {@see useObserverElement} hook. This hook
+ * is meant to be used for delayed element visibility callback.
+ *
+ * @param {IntersectionObserverInit} options
+ * @param {(element:Element)=>void} onElementVisible
+ * @param {string} observerId the name of the observer
+ * @param {string} elementIdKey the key of the unique identifier of each element.
+ * @param {string} delayForMS the number of milliseconds after which the observer should trigger the callback
+ */
+const useGlobalIntersectionObserverWithDelay = (
+  { threshold = 1, root = null, rootMargin = '0%' }: IntersectionObserverInit,
+  onElementVisible: (element: Element) => void,
+  observerId: string,
+  elementIdKey: string,
+  delayForMS: number,
+) => {
+  /**
+   * An object that will hold a reference of each unique identifier and
+   * the ID of the timeout function that will trigger when the element
+   * is visible for delayForMS number of seconds.
+   */
+  const timeouts = useRef<Record<any, number>>({});
+  const updateEntry = useCallback(
+    (entries: IntersectionObserverEntry[]): void => {
+      entries.forEach((visibleEntry) => {
+        let elementIntersectionDebuggerNode: HTMLElement;
+        if (isDev) {
+          const nodes = visibleEntry.target.getElementsByClassName('debugger');
+          // if at least one debugger node was found (should always be there but just to be extra sure)
+          if (nodes.length) {
+            elementIntersectionDebuggerNode = nodes[0] as HTMLElement;
+          }
+        }
+        const targetNode = visibleEntry.target as HTMLElement;
+        const elementId = targetNode.dataset[elementIdKey];
+        // if the element content container is interacting
+        if (visibleEntry.isIntersecting) {
+          setIntersectionDebuggerNodeText(
+            `Element is fully visible, triggering callback in ${delayForMS / 1000} seconds...`,
+            elementIntersectionDebuggerNode,
+          );
+          // Call the API after delayForMS
+          timeouts.current = {
+            ...timeouts.current,
+            [elementId]: window.setTimeout(() => {
+              onElementVisible(visibleEntry.target);
+              setIntersectionDebuggerNodeText(
+                `Element callback triggered`,
+                elementIntersectionDebuggerNode,
+              );
+            }, delayForMS),
+          };
+        } else {
+          // if it's no longer intersecting, clear the timeout
+          clearTimeout(timeouts.current[elementId]);
+          /**
+           * Delete the timeout function's ID from the global array since it's no longer
+           * needed and next time if the same element intersects, a new ID will be generated
+           * for that element.
+           */
+          const tempTimeouts = { ...timeouts.current };
+          delete tempTimeouts[elementId];
```

Comment:
ditto: `delete` keyword has bad performance. change to `timeouts.delete()` after map refactor.